Add note about serialization to update_columns documentation [ci skip]

RochesterinNYC opened the pull request:
Added detail to the update_columns documentation to clarify some of the confusion related to serialization as evident by #18037

rafaelfranca commented:
r? @sgrif

sgrif commented:
This is incorrect. The behavior of `update_columns` falls under the same category as [update_all](http://api.rubyonrails.org/classes/ActiveRecord/Relation.html#method-i-update_all)

sgrif commented:
Ugh hold on, let me see if I can straighten this out. There's still a mismatch here.

sgrif commented:
The documentation for `update_all` is incorrect. Can you have this method link to the docs for `update_all`, and change `update_all` to clarify that serialization occurs, but casting does not (e.g. you can pass it `model.attributes`, but you could not pass it `params`)

sgrif commented:
https://github.com/rails/rails/blob/master/activerecord/lib/active_record/relation.rb#L350-L351

> Values passed to #update_all will not go through Active Record's type-casting behavior.

https://github.com/rails/rails/blob/5d1d5f2840cd9aabd87e9dd4a110578d6c30db47/activerecord/test/cases/relation_test.rb#L300

`def test_update_all_goes_through_normal_type_casting`

<_<

RochesterinNYC commented:
Hm okay, I'm a little confused by the name of that test though, should it be `test_update_all_goes_through_normal_serialization` ? A quick read of the test seems like it's 

What's the canonical difference in rails between type-casting, casting, and serialization?

RochesterinNYC commented:
@sgrif Need a bit of clarification here to proceed. Pre-emptive thanks for your patience!

sgrif commented:
casting = user input -> rubyland
serialization = rubyland -> database

RochesterinNYC commented:
@sgrif Ah, gotcha. Thanks for the clarification.

Updated/cleaned up the PR accordingly.

sgrif commented on activerecord/lib/active_record/persistence.rb:
```diff
@@ -286,21 +286,9 @@ def update_column(name, value)
-    #
-    # * \Validations are skipped.
-    # * \Callbacks are skipped.
-    # * +updated_at+/+updated_on+ are not updated.
```
I think we should keep most of this, but add a note to this list that reads something along the lines of "attributes are serialized using the same rules as update_all"

sgrif commented on activerecord/lib/active_record/relation.rb:
```diff
@@ -346,10 +346,16 @@ def scoping
+    # * \Validations are skipped.
+    # * \Callbacks are skipped.
+    # * +updated_at+/+updated_on+ are not updated.
+    # * Involved models are not instantiated
```
This shouldn't move.

sgrif commented on activerecord/lib/active_record/relation.rb:
```diff
@@ -346,10 +346,16 @@ def scoping
+    # * +updated_at+/+updated_on+ are not updated.
+    # * Involved models are not instantiated
+    #
+    # However, values passed to #update_all will still go through Active Record's type-casting behavior.
```
This line makes me think it's the same as `.attributes = params && .save`, I think we should be more specific.

RochesterinNYC committed:
Clean up and correct documentation for update_columns and update_all [ci skip]

RochesterinNYC commented:
@sgrif Okay, followed your advice and just added an update_all reference bullet in `update_columns` and corrected and specified the behavior for `update_all`. I finally see that `sanitize_sql_hash_for_assignment` in the call stack for `update_all` is what ends up calling `serialize` on the attribute types.

RochesterinNYC commented:
Thanks!
